### PR TITLE
Refactor `launch.frontend` file loading

### PR DESCRIPTION
**Pull request by ivanpauno**
Changed the code a little bit, for better error handling and code reuse.
Aiming to use it for supporting `.launch.xml` and `.launch.yaml` with `ros2launch` command.

### CODE REVIEW

File: launch/launch/launch_description_sources/any_launch_file_utilities.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    ...
+
+
+def get_launch_description_from_any_launch_file(
+    launch_file_path: Text,
+    *,
+    parser: Type[Parser] = Parser
+) -> LaunchDescription:
+    """Load a given launch file (by path), and return the launch description from it."""
```
Comment (hidmic) [inline launch/launch/launch_description_sources/any_launch_file_utilities.py]: @ivanpauno consider documenting this function failure modes.

File: launch/launch/launch_description_sources/any_launch_file_utilities.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        return get_launch_description_from_frontend_launch_file(launch_file_path)
+    try:
+        return get_launch_description_from_python_launch_file(launch_file_path)
+    except (InvalidPythonLaunchFileError, SyntaxError):
+        pass
+    try:
+        return get_launch_description_from_frontend_launch_file(launch_file_path)
+    except InvalidFrontendLaunchFileError:
+        pass
+    raise InvalidLaunchFileError('Failed to load launch file')
```
Comment (hidmic) [inline launch/launch/launch_description_sources/any_launch_file_utilities.py]: @ivanpauno I'm somewhat inclined towards an `ask for forgiveness, not for permission` approach, specially if eventually   we stop relying (solely) on extensions to tell launch file types aparts.

TL;DR let's just attempt to load the file and deal with the exception if any.

File: launch/launch/frontend/parser.py [hunk 3]
```diff
@@ -94,14 +122,10 @@ def load(
         """Return an entity loaded with a markup file."""
         cls.load_parser_implementations()
```
Comment (hidmic) [inline launch/launch/frontend/parser.py]: @ivanpauno is this still necessary?

File: launch/launch/frontend/parser.py [hunk 4]
```diff
@@ -37,6 +38,12 @@
 }
 
 
+class InvalidFrontendLaunchFileError(Exception):
+    """Exception raised when the given Frontend launch file is not valid."""
```
Comment (hidmic) [inline launch/launch/frontend/parser.py]: @ivanpauno consider
```suggestion
    """Exception raised when the given frontend launch file is not valid."""
```

File: launch/launch/launch_description_sources/any_launch_file_utilities.py (hunk 2 above)
Comment (ivanpauno) [reply launch/launch/launch_description_sources/any_launch_file_utilities.py]: If I check that the extension is `.py` and the file have a syntax error, a nice `SyntaxError` will be raised.
If I just try load every possible implementation and then raise a generic error, it's not that clear.

I'm not completely relying on file extensions, just using it as a hint (if it's there, it's used).

I could try to do an "smarter" try load that prints nicer errors. What it's not easy, it will probably show confusing error sometimes.

File: launch/launch/frontend/parser.py (hunk 3 above)
Comment (ivanpauno) [reply launch/launch/frontend/parser.py]: Yes

File: launch/launch/frontend/parser.py [hunk 5]
```diff
@@ -86,28 +94,64 @@ def parse_description(self, entity: Entity) -> LaunchDescription:
...
+            raise RuntimeError('Not recognized frontend implementation')
+
     @classmethod
     def load(
         cls,
         file: Union[str, io.TextIOBase],
     ) -> (Entity, 'Parser'):
-        """Return an entity loaded with a markup file."""
+        """
+        Return an entity loaded with a markup file.
```
Comment (hidmic) [inline launch/launch/frontend/parser.py]: @ivanpauno consider
```suggestion
        Parse an Entity from a markup language-based launch file.
```

File: launch/launch/frontend/parser.py [hunk 6]
```diff
@@ -86,28 +94,64 @@ def parse_description(self, entity: Entity) -> LaunchDescription:
...
+
+        It first try the implementation specified by the extension (if it has).
```
Comment (hidmic) [inline launch/launch/frontend/parser.py]: @ivanpauno consider
```suggestion
       Parsers are exposed and provided by available frontend implementations. 
       To choose the right parser, it'll first attempt to infer the launch description format based on the filename extension, if any.
```

File: launch/launch/frontend/parser.py [hunk 7]
```diff
@@ -86,28 +94,64 @@ def parse_description(self, entity: Entity) -> LaunchDescription:
...
+        Then, it tries every possible implementation.
```
Comment (hidmic) [inline launch/launch/frontend/parser.py]: @ivanpauno consider
```suggestion
        If format inference fails, it'll try all available parsers one after the other.
```

File: launch/launch/frontend/parser.py [hunk 8]
```diff
@@ -86,28 +94,64 @@ def parse_description(self, entity: Entity) -> LaunchDescription:
...
-            if frontend_name in cls.frontend_parsers:
-                return cls.frontend_parsers[frontend_name].load(file)
-        # If not, apply brute force.
-        # TODO(ivanpauno): Maybe, we want to force correct file naming.
-        # In that case, we should raise an error here.
-        # TODO(ivanpauno): Recognize a wrong formatted file error from
-        # unknown front-end implementation error.
-        for implementation in cls.frontend_parsers.values():
+            try:
+                extension = file.rsplit('.', 1)[1]
```
Comment (hidmic) [inline launch/launch/frontend/parser.py]: @ivanpauno consider using `os.path.splitext` instead.

File: launch/launch/frontend/parser.py [hunk 9]
```diff
@@ -86,28 +94,64 @@ def parse_description(self, entity: Entity) -> LaunchDescription:
...
+                continue
             try:
                 return implementation.load(file)
             except Exception:
                 if is_a(file, io.TextIOBase):
                     file.seek(0)
-        raise RuntimeError('Not recognized front-end implementation.')
+        if ex_to_show is None:
+            raise InvalidFrontendLaunchFileError(
+                'The launch file may have a syntax error, or the frontend implementation '
```
Comment (hidmic) [inline launch/launch/frontend/parser.py]: @ivanpauno consider
```suggestion
                'The launch file may have a syntax error or its format is unknown'
```

File: launch/launch/launch_description_sources/any_launch_file_utilities.py [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    """
+    Load a given launch file (by path), and return the launch description from it.
+
+    :raise `InvalidLaunchFileError`: Failed to load launch file.
+        It's only showed with launch files without extension (or not recognized extensions).
+    :raise `SyntaxError`: Invalid file. The file may have a syntax error in it.
+    :raise `ValueError`: Invalid file. The file may not be a text file.
+    """
+    try:
+        extension = launch_file_path.rsplit('.', 1)[1]
```
Comment (hidmic) [inline launch/launch/launch_description_sources/any_launch_file_utilities.py]: @ivanpauno same about using `os.path.splitext`

File: launch/launch/launch_description_sources/any_launch_file_utilities.py [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    except IndexError:
+        extension = None
+    ex_to_show = None
+    tried_frontend = False
+    if extension == 'py':
+        try:
+            return get_launch_description_from_python_launch_file(launch_file_path)
+        except Exception as ex:
+            ex_to_show = ex
+    elif Parser.is_extension_valid(extension):
```
Comment (hidmic) [inline launch/launch/launch_description_sources/any_launch_file_utilities.py]: @ivanpauno do we need to check, as opposed to just calling `get_launch_description_from_frontend_launch_file(...)`?

File: launch/launch/launch_description_sources/frontend_launch_file_utilities.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+# Re-export name
+InvalidFrontendLaunchFileError = InvalidFrontendLaunchFileError
+
+
+def get_launch_description_from_frontend_launch_file(
+    frontend_launch_file_path: Text,
+    *,
+    parser: Type[Parser] = Parser
+) -> LaunchDescription:
+    """Load a given Frontend launch file (by path), and return the launch description from it."""
```
Comment (hidmic) [inline launch/launch/launch_description_sources/frontend_launch_file_utilities.py]: @ivanpauno consider
```suggestion
    """Load a `LaunchDescription` from a declarative (markup based) launch file."""
```

**Comment by ivanpauno** [comment]
@hidmic I think I have addressed your feedback better this time.

File: launch/launch/frontend/parser.py [hunk 13]
```diff
@@ -86,28 +97,68 @@ def parse_description(self, entity: Entity) -> LaunchDescription:
...
+        # Imported here, to avoid recursive import.
         cls.load_parser_implementations()
+
+        def get_key(extension):
+            def key(x):
+                return x[0] != extension
+            return key
+        exceptions = []
+        extension = ''
         if is_a(file, str):
```
Comment (hidmic) [inline launch/launch/frontend/parser.py]: @ivanpauno fyi: you may also get the extension if its a `file` object read from the filesystem.

File: launch/launch/launch_description_sources/frontend_launch_file_utilities.py (hunk 12 above)
Comment (hidmic) [reply launch/launch/launch_description_sources/frontend_launch_file_utilities.py]: @ivanpauno what happened with this?

**Comment by ivanpauno** [comment]
Rebased the PR with master, since there were merge conflicts.

CI (up to launch, launch_xml, launch_yaml, launch_ros, test_launch_ros, only fastrtps):
* Linux [Build Status](https://ci.ros2.org/job/ci_linux/7650/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/3761/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6257/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/7494/)

**Comment by ivanpauno** [comment]
Error is happening in the nightlies too. Merging!
